### DISCUSSION TITLE
Revisit nested modules

### DISCUSSION THREAD

**Discussion by konnov** [informalsystems/quint#496] [General]
As discussed in #483, the [language manual]() gives the following example:

```
module Outer {
  var x: int

  def xPlus(z) = x + z

  module Inner {
    val x2 = x + x
  }

  val inv =
    (Inner.x2 - x == x)
}
```

I would like to start a discussion on the need for nested modules. There are two immediate questions:

 - Do we need nested modules? It looks like we introduced them to have some form of namespaces. It would be good to see how other languages are using nested modules, if at all.
 - If we need nested modules, should we allow them to refer to the names of the outer modules, see #483.

/cc @lasarojc, @shonfeder, @bugarela, @thpani, @Kukovec

**Comment by bugarela** [comment]
This is actually the first time I'm really thinking about the possibility of removing nested modules. And I only now realize how much complicated module handling is only because of nested modules! See https://github.com/informalsystems/quint/issues/123. This code pattern is repeated in so many places and I'd have to introduce it in the language server as well somehow in order to properly support modules.

I'd say, considering how much immature is our current module/instances/imports support today, it would make more sense to remove nested modules and make things consistent without them, unless we are absolutely sure we want to support those. Now that I thought about how much work it is to maintain them, I don't think it is worth it.

Note: If we decide to remove them, it will be crucial that we either support multiple module definitions on a single file (right now the parser only parses the first module it encounters and its sub-modules), or multiple files. Otherwise, people will be able to define only a single module.

**Reply by lasarojc** [reply]
My attempt at using inner modules was exploratory and not based on a concrete need. For all the uses I can think of, multiple modules in a single file, as proposed by @bugarela, should work well enough.

**Comment by bugarela** [comment]
After our weekly meeting, here's how we want to proceed:

1. Remove support for nested module definition
2. Add support for namespaced module names (i.e. `module Foo::Bar {}`) (we actually already support this, but I'm adding here as it was proposed in the meeting)
3. Support multiple modules in a single file (see #547)
4. Eventually, we might introduce some new feature to provide a proper way of hiding.

I'll start removing support for nested modules immediately, and then work on the remaining module/import/instance issues.
